File: problems/678-valid-parenthesis-string.ts
function checkValidString(s: string): boolean {
  const leftParens: number[] = []
  const star: number[] = []

  for (let i = 0; i < s.length; i++) {
    const c = s[i]
    switch (c) {
      case "(":
        leftParens.push(i)
        break
      case ")":
        if (leftParens.length) {
          leftParens.pop()
        } else if (star.length) {
          star.pop()
        } else {
          return false
        }
        break
      case "*":
        star.push(i)
        break
    }
  }

  while (leftParens.length) {
    const index = leftParens.pop()!
    while (true) {
      if (!star.length) {
        return false
      }
      if (star.pop()! > index) {
        break
      }
    }
  }

  return true
}

checkValidString("(((((*)))**")

File: problems/678_valid_parenthesis_string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        left_parens: list[int] = []
        stars: list[int] = []

        for i, c in enumerate(s):
            match c:
                case "(":
                    left_parens.append(i)
                case ")":
                    if left_parens:
                        left_parens.pop()
                    elif stars:
                        stars.pop()
                    else:
                        return False
                case "*":
                    stars.append(i)

        while left_parens:
            i = left_parens.pop()
            while True:
                if not stars:
                    return False
                if stars.pop() > i:
                    break

        return True
